### ISSUE TITLE
[Bug] Can't replace ROM file while running [sf#564]

### ISSUE THREAD

**Issue by openMSX-import** [openMSX/openMSX#955]
**Reported by manuelbi on 2015-01-05 13:23 UTC**
When developing MSX software (e.g. a ROM), it would be useful to keep openMSX running and having the ROM replaced (by the compiler/assembler) without having to restart (or eject/reinsert) the ROM file. This is currently not possible.

On Linux, I did an experiment, replacing a Goonies ROM with a Gradius 1 ROM (with the same filename, just overwriting it while openMSX was running THe Goonies). In most cases, the MSX reset but nothing happened anymore (wrong mapper type).

But in one case I even got a bus error:

```
#0  0x00000000004bb40a in openmsx::CPUCore<openmsx::Z80TYPE>::executeInstructions (this=0x18f4630) at src/cpu/CPUCore.cc:1186
1186    CASE(C9) { int c = ret();         NEXT; }
(gdb) bt
#0  0x00000000004bb40a in openmsx::CPUCore<openmsx::Z80TYPE>::executeInstructions (this=0x18f4630) at src/cpu/CPUCore.cc:1186
#1  0x00000000004d2131 in openmsx::CPUCore<openmsx::Z80TYPE>::execute2 (
    this=this@entry=0x18f4630, fastForward=<optimized out>)
    at src/cpu/CPUCore.cc:2559
#2  0x00000000004d242e in openmsx::CPUCore<openmsx::Z80TYPE>::execute (
    this=0x18f4630, fastForward=<optimized out>) at src/cpu/CPUCore.cc:2516
#3  0x000000000050f12a in openmsx::MSXCPU::execute (this=<optimized out>, 
    fastForward=fastForward@entry=false) at src/cpu/MSXCPU.cc:153
#4  0x0000000000430271 in openmsx::MSXMotherBoard::execute (
    this=<optimized out>) at src/MSXMotherBoard.cc:477
#5  0x000000000044a331 in openmsx::Reactor::run (
    this=this@entry=0x7fff11284ba0, parser=...) at src/Reactor.cc:538
#6  0x0000000000410eaa in openmsx::main (argc=<optimized out>, 
    argv=<optimized out>) at src/main.cc:152
#7  0x0000000000411139 in main (argc=<optimized out>, argv=<optimized out>)
    at src/main.cc:181
```

I now tested with a similar 32kB ROM and I got another Bus error:

```
Program terminated with signal SIGBUS, Bus error.
#0  0x00000000004b641b in RDMEM_impl2<false, false> (cc=<optimized out>, 
    address=<optimized out>, this=<optimized out>) at src/cpu/CPUCore.cc:615
615         return line[address];
(gdb) bt
#0  0x00000000004b641b in RDMEM_impl2<false, false> (cc=<optimized out>, 
    address=<optimized out>, this=<optimized out>) at src/cpu/CPUCore.cc:615
#1  RDMEM_impl<false, false> (cc=<optimized out>, address=<optimized out>, 
    this=<optimized out>) at src/cpu/CPUCore.cc:625
#2  RDMEM_OPCODE (cc=<optimized out>, this=<optimized out>)
    at src/cpu/CPUCore.cc:631
#3  openmsx::CPUCore<openmsx::Z80TYPE>::executeInstructions (this=0x2765720)
    at src/cpu/CPUCore.cc:949
#4  0x00000000004d1b3c in openmsx::CPUCore<openmsx::Z80TYPE>::executeSlow (
    this=this@entry=0x2765720) at src/cpu/CPUCore.cc:2501
#5  0x00000000004d2049 in openmsx::CPUCore<openmsx::Z80TYPE>::execute2 (
    this=this@entry=0x2765720, fastForward=<optimized out>)
    at src/cpu/CPUCore.cc:2551
#6  0x00000000004d242e in openmsx::CPUCore<openmsx::Z80TYPE>::execute (
    this=0x2765720, fastForward=<optimized out>) at src/cpu/CPUCore.cc:2516
#7  0x000000000050f12a in openmsx::MSXCPU::execute (this=<optimized out>, 
    fastForward=fastForward@entry=false) at src/cpu/MSXCPU.cc:153
#8  0x0000000000430271 in openmsx::MSXMotherBoard::execute (
    this=<optimized out>) at src/MSXMotherBoard.cc:477
#9  0x000000000044a331 in openmsx::Reactor::run (
    this=this@entry=0x7fff75497bc0, parser=...) at src/Reactor.cc:538
#10 0x0000000000410eaa in openmsx::main (argc=<optimized out>, 
    argv=<optimized out>) at src/main.cc:152
```

Tried another time, and now it worked without Bus error.

On Windows, I hear it is not working at all. I guess Windows says the file is in use and cannot be overwritten.

**Comment by openMSX-import** [comment]
**Commented by manuelbi on 2015-01-05 13:30 UTC**
- Description has changed:

Diff:

```

--- old
+++ new
@@ -3,6 +3,8 @@
 On Linux, I did an experiment, replacing a Goonies ROM with a Gradius 1 ROM (with the same filename, just overwriting it while openMSX was running THe Goonies). In most cases, the MSX reset but nothing happened anymore (wrong mapper type).

 But in one case I even got a bus error:
+
+~~~~
 #0  0x00000000004bb40a in openmsx::CPUCore<openmsx::Z80TYPE>::executeInstructions (this=0x18f4630) at src/cpu/CPUCore.cc:1186
 1186   CASE(C9) { int c = ret();         NEXT; }
 (gdb) bt
@@ -22,10 +24,11 @@
     argv=<optimized out>) at src/main.cc:152
 #7  0x0000000000411139 in main (argc=<optimized out>, argv=<optimized out>)
     at src/main.cc:181
-
+~~~~

 I now tested with a similar 32kB ROM and I got another Bus error:

+~~~~
 Program terminated with signal SIGBUS, Bus error.
 #0  0x00000000004b641b in RDMEM_impl2<false, false> (cc=<optimized out>, 
     address=<optimized out>, this=<optimized out>) at src/cpu/CPUCore.cc:615
@@ -54,7 +57,7 @@
     this=this@entry=0x7fff75497bc0, parser=...) at src/Reactor.cc:538
 #10 0x0000000000410eaa in openmsx::main (argc=<optimized out>, 
     argv=<optimized out>) at src/main.cc:152
-
+~~~~

 Tried another time, and now it worked without Bus error.


```

**Comment by openMSX-import** [comment]
**Updated by manuelbi on 2015-01-05 13:31 UTC**
- **labels**:  --> development

**Comment by MBilderbeek** [comment]
I tried this on Linux, and there it works fine now (probably since years ago already). I could even edit the ROM file and next time the code in the ROM file used the data, it displayed the updated data :)

However, on Windows, this doesn't work, the ROM file is not allowed to be written.

**Comment by MBilderbeek** [comment]
To open a file, we use OS specific code. We use memory  mapped files. And on Windows use use this stuff (see `src/file/LocalFile.cc`):
```
std::span<const uint8_t> LocalFile::mmap()
{
	size_t size = getSize();
	if (size == 0) return {static_cast<uint8_t*>(nullptr), size};

	if (!mmem) {
		int fd = _fileno(file.get());
		if (fd == -1) {
			throw FileException("_fileno failed");
		}
		auto hFile = reinterpret_cast<HANDLE>(_get_osfhandle(fd)); // No need to close
		if (hFile == INVALID_HANDLE_VALUE) {
			throw FileException("_get_osfhandle failed");
		}
		assert(!hMmap);
		hMmap = CreateFileMapping(hFile, nullptr, PAGE_WRITECOPY, 0, 0, nullptr);
		if (!hMmap) {
			throw FileException(
				"CreateFileMapping failed: ", GetLastError());
		}
		mmem = static_cast<uint8_t*>(MapViewOfFile(hMmap, FILE_MAP_COPY, 0, 0, 0));
		if (!mmem) {
			DWORD gle = GetLastError();
			CloseHandle(hMmap);
			hMmap = nullptr;
			throw FileException("MapViewOfFile failed: ", gle);
		}
	}
	return {mmem, size};
}
```
Perhaps these are not the right options for these API calls? Or should we use a different combination to allow processes to still write the memory mapped file on Windows?
We need a Windows specialist to step up here.... Anyone?

**Comment by mthuurne** [comment]
Before diving into mmap, is it even possible for two processes to have the same file open simultaneously in Windows? By default it is not (unlike POSIX), but maybe there are options to allow it.

**Comment by MBilderbeek** [comment]
Note that I understood from Roman that in blueMSX, it is possible to have the emulator running with the ROM file and then overwrite it and get the updated ROM file without restarting blueMSX. You do need to reset the MSX though. Maybe it's simply reloading the ROM files each MSX reset or something?

**Comment by m9710797** [comment]
I haven't checked the blueMSX source code, but very likely it'll indeed just read the ROM file in memory and then close the file.

It's easy to do the same in openMSX, but then we loose the advantage of mmap'ing the file. But now, 20 years later, that's maybe not so important anymore?

Note that, in openMSX, resetting the MSX won't be sufficient to reload the ROM. Instead you really need to eject/reinsert the cartridge or recreate the MSX machine instance.

**Comment by MBilderbeek** [comment]
@m9710797 I'm not so sure about that. At least on Linux, mmap now provides a very convenient feature: we can modify the ROM file on the spot, and it's immediately visible in the emulated MSX. I think that's a very good thing to keep.
So, ideally, we should realize the same on Windows. And for that there are 2 things: 1) find out whether it's possible to make a memory map whilst not blocking the file for writing by another process and 2) get the proper implementation if it is.

The reset thing is what seems to happen in blueMSX. It's just to explain why it (partly) works there. So to be clear: on non-Windows blueMSX is more limited than openMSX. But on Windows, openMSX is more limited, as the ROM file cannot be overwritten at all when openMSX has it open. And in blueMSX you must apparently first reset. I'm not sure if it reopens on reset, but that's just a possible explanation of the behaviour in blueMSX.

**Comment by m9710797** [comment]
About the Linux behavior: you say "it's a good thing to keep". But do you know of anyone who actually uses this feature in practice?
We (or at least I) never consciously decided to make it work like this. It was merely a tweak to speedup startup time (don't have to wait till the full file is loaded in memory). The "changes to the file are immediately visible in openMSX" is at most an accidental feature. And I also doubt it'll work correctly in all cases. Here are two examples:

- From the mmap man page: 
    MAP_PRIVATE
    Create a private copy-on-write mapping.  Updates to the mapping are not  visible  to  other  pro‐
    cesses mapping the same file, and are not carried through to the underlying file.  It is unspeci‐
    fied whether changes made to the file after the mmap() call are visible in the mapped region.

    In other words: it may or may not be the case that changes to the file will be visible in openMSX. For example it could be that (with page granularity, typically 4kB) pages that have not been read yet by openMSX (nor by the OS read-ahead) will be taken from the file. But pages that have already been read before will not be updated in openMSX memory after the corresponding part of the ROM file has been updated.

- We cache the calculation of the sha1sum of the ROM file (for savestates, e.g. periodic snapshots for reverse). This cache won't be invalidated when the underlying ROM file changes.

**Comment by MBilderbeek** [comment]
Okay thanks for that new info...

Most important thing is that we support this use case on all platforms: it must be possible to overwrite the file that is used in the emulation. And on top of that, to keep on par with blueMSX, with having to reinsert the ROM file.

Can someone help figure out how blueMSX achieves this? I mean, especially the last part... (No need to reinsert.)

**Comment by m9710797** [comment]
No need to check how blueMSX achieves this. I'm 99% sure it'll read the full ROM file in memory, and re-read it on reset.
But I'm not convinced we need to exactly replicate this behavior. More specifically I mean the forced re-read on reset.

Instead what I'd like to "promise" is:
* **NO** guarantees what happens when you change the content of the ROM file while openMSX is using it:
  * OpenMSX may continue using the old content, or the new content, or even a mix between the two.
* The only way to force openMSX to re-read the ROM files is to eject and reinsert the ROM file (or recreate the machine in case of system ROMs). In other words: reset gives no such guarantee.

An alternative could have been to "promise" that openMSX will continue to use the old ROM content until explicitly requested to refresh. At first sight it seems we can accomplish this by not using mmap() (instead read the full file at startup). But that is still insufficient for the following scenario:
* Reverse: a backwards step in time will recreate the machine and thus read the new content of ROM files.
* For a forward step it depends on the size of the step and the position of snapshots: if we cross a snapshot, we'll recreate the machine and use the new content, if not then we'll only fast-forward the current machine and thus continue using the current ROM content.

But these are very much internal implementation details. I don't want to make promises based on that (also to keep the freedom to change the implementation in the future). So therefor I'd like to give no such promises.

The only thing (IMHO) we may want to change is to not lock the ROM files for writing by external applications. That means on Linux we don't need to do anything. And on windows we drop the mmap() code (ok, it's called differently on windows). Either only for ROM files or in general. (And to reduce the amount of platform specific code, we then also might want to drop mmap() for ROM files on Linux).

**Comment by MBilderbeek** [comment]
@m9710797 FYI: what I hear from a Mac user is that despite that we're using mmap, changes to the file are *not* reflected to the MSX immediately there.

**Comment by m9710797** [comment]
@MBilderbeek that's as it should be, and we have the same on Linux. We use mmap() with the MAP_PRIVATE flag. From the man page:

      MAP_PRIVATE
              Create a private copy-on-write mapping.  Updates to the mapping are not visible to other  pro‐
              cesses  mapping  the  same  file,  and  are not carried through to the underlying file.  It is
              unspecified whether changes made to the file after the mmap() call are visible in  the  mapped
              region.

In other words, we only use mmap() as a faster version of reading the file. That is the mmap() system call returns immediately, and then in the background the OS will populate the mapping with data from the file on disk. But when exactly those reads happen in unspecified (it happens at some point before the program reads from that memory location, possibly blocking the program if the OS did not have time to populate it). Once (a page of) the mapping has been populated it will not be read again from disk (Not 100% sure about this part: maybe unless memory is tight and the mapping has been swapped out, then it may get re-populated from the file instead of swapping-out and swapping-in again (we can skip the swap-out part)).

Recently we also added the MAP_POPULATE flag to mmap(). This is a hint, it asks the OS to start populating the mapping as soon at the mmap() call is made (and not e.g. only when the first page of the mapping is accessed).

**Comment by MBilderbeek** [comment]
Well, for me it still works: e.g. I start up openMSX with the ROM of Alpha Roid, change a character on the title screen, start the game and get it to game over and then the title screen shows the modified character.

Anyways, you had a good workaround/idea to reload the ROMs that seems to work (at least on mac): `reverse goto [dict get [reverse status] current]`
